modin/core/dataframe/algebra/default2pandas/groupby.py

```diff
@@ -270,6 +293,172 @@ def build_groupby(cls, func):
+
+        Examples
+        --------
+        >>> groupby_result = compute_groupby_without_processing_as_index_parameter()
+        >>> if not as_index:
+        >>>     drop, lvls_to_drop, cols_to_drop = handle_as_index(**extract_required_params(groupby_result))
+        >>>     if len(lvls_to_drop) > 0:
+        >>>         groupby_result.index = groupby_result.index.droplevel(lvls_to_drop)
+        >>>     if len(cols_to_drop) > 0:
+        >>>         groupby_result = groupby_result.drop(columns=cols_to_drop)
+        >>>     groupby_result_with_processed_as_index_parameter = groupby_result.reset_index(drop=drop)
+        >>> else:
+        >>>     groupby_result_with_processed_as_index_parameter = groupby_result
+        """
+        # If the method is "size" then the result contains only one unique named column
+        # and we don't have to worry about any naming conflicts, so inserting all of
+        # the "by" into the result (just a fast-path)
+        if method == "size":
+            return False, [], []
+
+        if is_multi_by is None:
+            is_multi_by = len(internal_by_cols) > 1
+
+        # Pandas logic of resolving naming conflicts is the following:
+        #   1. If any categorical is in 'by' and 'by' is multi-column, then the categorical
+        #      index is prioritized: drop intersected columns and insert all of the 'by' index
+        #      levels to the frame as columns.
+        #   2. Otherwise, aggregation result is prioritized: drop intersected index levels and
+        #      insert the filtered ones to the frame as columns.
+        if by_cols_dtypes is not None:
+            keep_index_levels = (
+                is_multi_by
+                and selection is None
+                and any(isinstance(x, pandas.CategoricalDtype) for x in by_cols_dtypes)
+            )
+        else:
+            keep_index_levels = False
+
+        # 1. We insert 'by'-columns to the result at the beginning of the frame and so only to the
+        #    first partition, if partition_idx != 0 we just drop the index. If there are no columns
+        #    that are required to drop (keep_index_levels is True) then we can exit here.
+        # 2. We don't insert 'by'-columns to the result if 'by'-data came from a different
+        #    frame (drop is False), there's only one exception for this rule: if the `method` is "size",
+        #    so if (drop is False) and method is not "size" we just drop the index and so can exit here.
+        if (not keep_index_levels and partition_idx != 0) or (
+            not drop and method != "size"
+        ):
+            return True, [], []
+
+        if not isinstance(internal_by_cols, pandas.Index):
+            if not is_list_like(internal_by_cols):
+                internal_by_cols = [internal_by_cols]
+            internal_by_cols = pandas.Index(internal_by_cols)
+
+        internal_by_cols = (
+            internal_by_cols[~internal_by_cols.str.match(r"__reduced__.*", na=False)]
```

Comment:
```suggestion
            internal_by_cols[~internal_by_cols.str.startswith("__reduced__", na=False)]
```
or something like that. Do not use regexps unless really necessary, in this case for performance reasons (however, some measurements would be good to prove me right or wrong tho)

Reply:
`startswith` is slightly faster indeed:
```
startswith:
        Total time of 10 runs: 6.242122787982225; Mean time: 0.6242122787982225
regexp:
        Total time of 10 runs: 7.681726414011791; Mean time: 0.7681726414011791
```

<details><summary>reproduced</summary>

```python
import pandas
import timeit
import random

random.seed(42)
n_elements = 1_000_000
n_runs = 10

reduced_columns_percentage = 0.1
almost_reduced_columns_percentage = 0.2
regular_columns_percentage = 1 - (
    reduced_columns_percentage + almost_reduced_columns_percentage
)

reduced = [
    f"__reduced__someth{i}ngsomething"
    for i in range(int(n_elements * reduced_columns_percentage))
]
almost_reduced = [
    f"__reduced_someth{i}ngsomething"
    for i in range(int(n_elements * almost_reduced_columns_percentage))
]
regular = [
    f"someth{i}ngsomething" for i in range(int(n_elements * regular_columns_percentage))
]

index = reduced + almost_reduced + regular
random.shuffle(index)

index = pandas.Index(index)

startswith_time = timeit.timeit(
    lambda: index.str.startswith("__reduced__", na=False), number=n_runs
)
regexp_time = timeit.timeit(
    lambda: index.str.match(r"__reduced__.*", na=False), number=n_runs
)

print(
    f"startswith:\n\tTotal time of {n_runs} runs: {startswith_time}; Mean time: {startswith_time / n_runs}"
)
print(f"regexp:\n\tTotal time of {n_runs} runs: {regexp_time}; Mean time: {regexp_time / n_runs}")
```

</details>

P.S. Timings of half a second for this filtering may scare, but note, that in the repro I filter a million values when in the actual code the filtered object is an Index, containing column names to group on, its length is unlikely to be big.